pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/ReferralSourceServiceTest.java

```diff
@@ -0,0 +1,347 @@
+        Assert.assertNotNull(questionsAnswers);
+        List<Map<String, Object>> qa = questionsAnswers.stream()
+                .filter(q -> !q.get(DDPActivityConstants.DDP_ACTIVITY_STABLE_ID).equals(
+                        DDPActivityConstants.ENROLLMENT_FIND_OUT)).collect(Collectors.toList());
+        enrollment.setQuestionsAnswers(qa);
+        log.info("Enrollment: " + enrollment);
+
+        String refSourceId = ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(lessActivities));
+        Assert.assertEquals("NA", refSourceId);
+    }
+
+    @Test
+    public void getRefSourceBadPossibleValues() throws Exception {
+        mockDao.refSourceValues = TestUtil.readFile("RefSourceFieldSettingsBadValues.json");
+
+        // this possible value is still intact
+        List<String> answer = List.of("FACEBOOK");
+        applyFoundOutAnswer(activities, answer);
+        String refSourceId = ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+        Assert.assertEquals("FACEBOOK", refSourceId);
+
+        // the NA is removed
+        answer = new ArrayList<>();
+        applyFoundOutAnswer(activities, answer);
+        try {
+            ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+            Assert.fail("convertReferralSources should throw when possible values NA is missing");
+        } catch (Exception e) {
+            Assert.assertTrue(e.getMessage().contains("does not include a 'NA' key"));
+        }
+    }
+
+    @Test
+    public void getRefSourceBadDetails() throws Exception {
+        mockDao.refSourceDetails = TestUtil.readFile("RefSourceFieldSettingsBadDetails.json");
+
+        // this answer is not a valid map key
+        List<String> answer = List.of("FACEBOOK");
+        applyFoundOutAnswer(activities, answer);
+        try {
+            ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+            Assert.fail("convertReferralSources should throw when there is no corresponding REF_SOURCE for "
+                    + "participant provided referral source");
+        } catch (Exception e) {
+            Assert.assertTrue(e.getMessage().contains("no corresponding REF_SOURCE"));
+        }
+
+        // this answer does not map to a valid REF_SOURCE
+        answer = List.of("TWITTER");
+        applyFoundOutAnswer(activities, answer);
+        try {
+            ReferralSourceService.convertReferralSources(ReferralSourceService.getReferralSources(activities));
+            Assert.fail("convertReferralSources should throw when participant provided referral source maps to "
+                    + "an invalid REF_SOURCE ID");
+        } catch (Exception e) {
+            Assert.assertTrue(e.getMessage().contains("Invalid REF_SOURCE ID"));
+        }
+    }
+
```

Comment:
Note to reviewer: the tests above this were reviewed in May, when they were part of a PR to support PEPPER-575.